core/src/opensearch-cluster/os-cluster.ts

```diff
@@ -0,0 +1,256 @@
+import { Construct } from 'constructs';
+import {
+  aws_iam,
+  aws_opensearchservice,
+  RemovalPolicy,
+  aws_secretsmanager,
+  custom_resources,
+  CustomResource,
+  aws_lambda_nodejs,
+  Stack,
+  CfnOutput,
+} from 'aws-cdk-lib';
+import { Architecture, Runtime } from 'aws-cdk-lib/aws-lambda';
+import { ManagedPolicy } from 'aws-cdk-lib/aws-iam';
+
+export interface OpensearchClusterProps {
+  readonly domainName?: string;
+  readonly accessRoles: aws_iam.Role[];
+  readonly adminUsername: string;
+  readonly usernames: string[];
+}
+
+export class OpensearchCluster extends Construct {
+  /**
+   * Get an existing OpensearchCluster based on the cluster name property or create a new one
+   * only one Opensearch cluster can exist per stack
+   * @param {Construct} scope the CDK scope used to search or create the cluster
+   * @param {OpensearchClusterProps} props the OpensearchClusterProps [properties]{@link OpensearchClusterProps} if created
+   */
+  public static getOrCreate(scope: Construct, props: OpensearchClusterProps) {
+    const stack = Stack.of(scope);
+    const id = props.domainName || OpensearchCluster.DEFAULT_DOMAIN_NAME;
+
+    let opensearchCluster: OpensearchCluster;
+
+    if (stack.node.tryFindChild(id) == undefined) {
+      opensearchCluster = new OpensearchCluster(stack, id, props);
+    }
+
+    return (stack.node.tryFindChild(id) as OpensearchCluster) || opensearchCluster!;
+  }
+
+  public static readonly DEFAULT_DOMAIN_NAME = 'opensearch-platform';
+  public readonly domainName: string;
+  public readonly domain: aws_opensearchservice.Domain;
+  private readonly apiFn: aws_lambda_nodejs.NodejsFunction;
+  private readonly masterRole: aws_iam.Role;
+  private prevCr?: CustomResource;
+
+  /**
+   * @public
+   * Constructs a new instance of the OpensearchCluster class
+   * @param {Construct} scope the Scope of the AWS CDK Construct
+   * @param {string} id the ID of the AWS CDK Construct
+   * @param {OpensearchClusterProps} props the OpensearchCluster [properties]{@link OpensearchClusterProps}
+   */
+
+  constructor(scope: Construct, id: string, props: OpensearchClusterProps) {
+    super(scope, id);
+    const { accessRoles, adminUsername, usernames } = props;
+
+    this.domainName = props.domainName ?? OpensearchCluster.DEFAULT_DOMAIN_NAME;
+
+    this.masterRole = new aws_iam.Role(this, 'AccessRole', {
```

Comment:
masterRole has very wide policies that need to be scoped down.
CloudWatchLogsFullAccess could be replaced by `service-role/AWSLambdaBasicExecutionRole`

AmazonOpenSearchServiceFullAccess policy could be replaced by:
```
            PolicyStatement(
                actions=['es:UpdateElasticsearchDomainConfig'],
                resources=['*'],
            ))
```
For private networks, need to add:
```
            PolicyStatement(
                actions=[
                    'ec2:CreateNetworkInterface',
                    'ec2:DescribeNetworkInterfaces',
                    'ec2:DeleteNetworkInterface',

                ],
                resources=['*'],
            ))
```